downstairs/src/repair.rs

```diff
@@ -0,0 +1,386 @@
+use dropshot::ConfigLogging;
+use dropshot::ConfigLoggingLevel;
+use dropshot::HttpError;
+use dropshot::HttpResponseOk;
+use dropshot::HttpServerStarter;
+use dropshot::RequestContext;
+use dropshot::{endpoint, Path};
+use http::{Response, StatusCode};
+use hyper::Body;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+
+use super::*;
+use crate::region::{extent_dir, extent_file_name, extent_path};
+
+/**
+ * Our context is the root of the region we want to serve.
+ */
+pub struct FileServerContext {
+    region_dir: PathBuf,
+}
+
+/**
+ * Build the API.  If requested, dump it to stdout.
+ * This allows us to use the resulting output to build the client side.
+ */
+pub fn build_api(
+    show: bool,
+) -> Result<ApiDescription<FileServerContext>, String> {
+    let mut api = ApiDescription::new();
+    api.register(get_extent).unwrap();
+    api.register(get_db).unwrap();
+    api.register(get_shm).unwrap();
+    api.register(get_wal).unwrap();
+    api.register(get_files_for_extent).unwrap();
+
+    if show {
+        api.openapi("downstairs-repair", "1")
+            .write(&mut std::io::stdout())
+            .map_err(|e| e.to_string())?;
+    }
+    Ok(api)
+}
+
+pub async fn repair_main(
+    ds: &Arc<Mutex<Downstairs>>,
+    addr: SocketAddr,
+) -> Result<(), String> {
+    /*
+     * We must specify a configuration with a bind address.
+     */
+    let config_dropshot = ConfigDropshot {
+        bind_address: addr,
+        request_body_max_bytes: 1024,
+        tls: None,
+    };
+
+    /*
+     * For simplicity, configure an "info"-level logger that writes to
+     * stderr assuming that it's a terminal.
+     */
+    let config_logging = ConfigLogging::StderrTerminal {
+        level: ConfigLoggingLevel::Info,
+    };
+    let log = config_logging
+        .to_logger("example-basic")
+        .map_err(|error| format!("failed to create logger: {}", error))?;
+
+    /*
+     * Build a description of the API
+     */
+    let api = build_api(false)?;
+
+    /*
+     * Record the region directory where all the extents and metadata
+     * files live.
+     */
+    let ds = ds.lock().await;
+    let region_dir = ds.region.dir.clone();
+    drop(ds);
+
+    let context = FileServerContext { region_dir };
+
+    println!("Repair listens on {}", addr);
+    /*
+     * Set up the server.
+     */
+    let server = HttpServerStarter::new(&config_dropshot, api, context, &log)
+        .map_err(|error| format!("failed to create server: {}", error))?
+        .start();
+
+    /*
+     * Wait for the server to stop.  Note that there's not any code to shut
+     * down this server, so we should never get past this point.
+     */
+    server.await
+}
+
+#[derive(Deserialize, JsonSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct Eid {
+    eid: u32,
+}
+
+/**
+ * Stream the contents of the data file for the given extent ID
+ */
+#[endpoint {
+    method = GET,
+    path = "/extent/{eid}/data",
```

Comment:
Am I reading this right that these 4 endpoints are basically the same? What would you think of something like this:

```rust
    #[derive(Deserialize, JsonSchema)]
    pub enum FileType {
        #[serde(rename = "data")
        Data,
        #[serde(rename = "db")
        Database,
        #[serde(rename = "db-shm")
        DatabaseSharedMemory, // ??
        #[serde(rename = "db-wal")
        DatabaseLog,
    }

    // ...

    struct FileSpec {
        eid: u32,
        file_type: FileType,
    }

    #[endpoint {
        method = GET,
        path = "/extent/{eid}/{file_type}",
    }]
    // ...
```

Reply:
Yeah, that's simpler.  I've updated it a bit and then updated a bit more now that their is an enum for `ExtentExtension`. I can probably polish it a bit more, but at some point I want to get this back under test.